src/devices/directv.c

```diff
@@ -0,0 +1,532 @@
+                        // Did we expect the EOR here or not, though?
+                        if (dtv_bit_idx < (DTV_BITLEN_MAX - 1)) {
+                            if (decoder->verbose > 1) {
+                                fprintf(stderr, "directv: bitbuffer contains unexpected bits after unexpected End Of Row bits.\n");
+                            }
+                            return 0;
+                        }
+                        if (decoder->verbose > 1) {
+                            fprintf(stderr, "directv: bitbuffer contains unexpected bits after the expected End Of Row bits.\n");
+                        }
+                        return 0;
+                    }
+                    // End of bitrow and EOR just had some trailing 0 bits.
+                    // Did we expect the EOR here or not, though?
+                    if (dtv_bit_idx < (DTV_BITLEN_MAX - 1)) {
+                        if (decoder->verbose > 1) {
+                            fprintf(stderr, "directv: bitbuffer contains unexpected End Of Row bits before we received complete data.\n");
+                        }
+                        return 0;
+                    }
+                    // Everything is good enough if we just ignore the trailing 0 bits.
+                    break;
+                }
+                // End of bitrow and either 1000 or 1100 received. 1000 is probably expected. 1100 is definitely bad.
+                if (dtv_twobits > 0) {
+                    // We shouldn't end bitrow with a 1100 data unit
+                    if (decoder->verbose > 1) {
+                        fprintf(stderr, "directv: bitbuffer contains unexpected 1100 bits at end of bitrow.\n");
+                    }
+                    return 0;
+                }
+                // 1000 is EOR and is expected at end of bitrow, but did we fill up dtv yet?
+                if (dtv_bit_idx < (DTV_BITLEN_MAX - 1)) {
+                    if (decoder->verbose > 1) {
+                        fprintf(stderr, "directv: bitbuffer contains End Of Row marker but not enough row data provided (%d < %d).\n", dtv_bit_idx, DTV_BITLEN_MAX );
+                    }
+                    return 0;
+                }
+                // Great!  We got the 1000 EOR exactly where we expected it.
+                break;
+            }
+            // will be either 100 or 110 at end of the bitrow here.  Not good.
+            if (decoder->verbose > 1) {
+                fprintf(stderr, "directv: bitbuffer contains unexpected %s bits at end of bitrow.\n", dtv_twobits==0?"100":"110");
+            }
+            return 0;
+        }
+        // will be either 10 or 11 at the end of bitrow here.  Not good.
+        if (decoder->verbose > 1) {
+            fprintf(stderr, "directv: bitbuffer contains unexpected %s bits at end of bitrow.\n", dtv_twobits==0?"10":"11");
+        }
+        return 0;
+    }
+    // We've received an acceptable decoded dtv value here.
```

Comment:
Oh, this was kinda messy. Lets hope the DMC decoder can be used to simplify the logic.

Reply:
I didn't check properly, kind-of-DMC rather as the bits are of variable length presumably. Something like "DPWM" -- if that makes sense. A new demod  derived from DMC/PWM should do.

Reply:
Good point.  After thinking about it this way (as some kind of differential PWM), I realized I was going about decoding all wrong.  I've completely rewritten the decoding by only looking at symbol transitions, making it so much more robust (works for symbol inversions) and generalized.